src/storage/src/upsert/types.rs

```diff
@@ -154,62 +172,94 @@ pub struct MergeValue<V> {
+/// to store it any way they want. During consolidation, the logic is more complex.
 /// See the docs on `StateValue::merge_update` for more information.
 ///
-///
 /// Note also that this type is designed to support _partial updates_. All values are
 /// associated with an _order key_ `O` that can be used to determine if a value existing in the
 /// `UpsertStateBackend` occurred before or after a value being considered for insertion.
 ///
 /// `O` typically required to be `: Default`, with the default value sorting below all others.
-/// Values consolidated during snapshotting consolidate correctly (as they are actual
+/// During consolidation, values consolidate correctly (as they are actual
 /// differential updates with diffs), so order keys are not required.
-#[derive(Clone, serde::Serialize, serde::Deserialize, Debug)]
-pub enum StateValue<O> {
-    Snapshotting(Snapshotting),
-    Value(Value<O>),
+#[derive(Clone, serde::Serialize, serde::Deserialize)]
+pub enum StateValue<T, O> {
+    Consolidating(Consolidating),
+    Value(Value<T, O>),
+}
+
+impl<T, O> std::fmt::Debug for StateValue<T, O> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            StateValue::Consolidating(_) => write!(f, "Consolidating"),
+            StateValue::Value(_) => write!(f, "Value"),
+        }
+    }
 }
 
 /// A totally consolidated value stored within the `UpsertStateBackend`.
 ///
-/// This type contains support for _tombstones_, that contain an _order key_.
+/// This type contains support for _tombstones_, that contain an _order key_,
+/// and provisional values.
+///
+/// What is considered finalized and provisional depends on the implementation
+/// of the UPSERT operator: it might consider everything that it writes to its
+/// state finalized, and assume that what it emits will be written down in the
+/// output exactly as presented. Or it might consider everything it writes down
+/// provisional, and only consider updates that it _knows_ to be persisted as
+/// finalized.
+///
+/// Provisional values should only be considered while still "working off"
+/// updates with the same timestamp at which the provisional update was
+/// recorded.
 #[derive(Clone, serde::Serialize, serde::Deserialize, Debug)]
-pub enum Value<O> {
-    Value(UpsertValue, O),
+pub enum Value<T, O> {
+    FinalizedValue(UpsertValue, O),
     Tombstone(O),
+    ProvisionalValue {
+        // We keep the finalized value around, because the provisional value is
+        // only valid when processing updates at the same timestamp. And at any
+        // point we might still require access to the finalized value.
+        finalized_value: Option<Box<(UpsertValue, O)>>,
+        // A provisional value of `None` is a provisional tombstone.
+        //
+        // WIP: We can also box this, to keep the size of StateValue as it was
+        // previously.
+        provisional_value: (Option<UpsertValue>, T, O),
+    },
 }
```

Comment:
Hm, it feels like this struct is trying to do two jobs at the same time. It makes much more sense to me to keep the original enum, which encodes the fact that a value can be either a value or a tombstone, and create a new struct that encodes that each key always has a final value and an optional provisional value:

```rust
enum Value<0> {
    Value(UpsertValue, O),
    Tombstone(O),
}

struct Value2<T, O> {
    final: Value<O>,
    provisional: Option<(Value<O>, Τ)>,
}
```

I believe this will probably dedup most of the methods that are written below since you can write the common parts for `Value<O>` once and then have a single method that gives you either the final Value<O> or the provisional one

Reply:
> and create a new struct that encodes that each key always has a final value and an optional provisional value

When ingesting the initial snapshot from the source, keys will have _no_ final value, and possibly a provisional value. An that provisional value itself could be a tombstone. :sweat_smile:

I'm very happy if we could simplify this thing, though! But with the above, what I have still feels necessary, no?